modules/pymol/shortcut.py

```diff
@@ -1,184 +1,171 @@
-
-        def append(self,kee):
-            self.keywords.append(kee)
-            self.add_one(kee)
-            self._rebuild_finalize()
-
-        def auto_err(self,kee,descrip=None):
-            result = None
-            if kee not in self.shortcut:
-                if descrip is not None:
-                    msg = "Error: unknown %s: '%s'." % (descrip, kee)
-                    lst = self.interpret('')
-                    if is_list(lst):
-                        if len(lst)<100:
-                            lst.sort()
-                            lst = parsing.list_to_str_list(lst)
-                            msg += " Choices:\n"
-                            msg += "\n".join(lst)
-                    raise parsing.QuietException(msg)
-
-            else:
-                result = self.interpret(kee)
-                if not is_string(result):
-                    if descrip is not None:
-                        lst = parsing.list_to_str_list(result)
-                        msg = "Error: ambiguous %s:\n%s" % (descrip, '\n'.join(lst))
-                        raise parsing.QuietException(msg)
+# A* -------------------------------------------------------------------
+# B* This file contains source code for the PyMOL computer program
+# C* Copyright (c) Schrodinger, LLC.
+# D* -------------------------------------------------------------------
+# E* It is unlawful to modify or remove this copyright notice.
+# F* -------------------------------------------------------------------
+# G* Please see the accompanying LICENSE file for further information.
+# H* -------------------------------------------------------------------
+# I* Additional authors of this source file include:
+# -*
+# -*
+# -*
+# Z* -------------------------------------------------------------------
+
+from typing import Iterable, Optional
+from collections import defaultdict
+from pymol import parsing
+
+
+class Shortcut:
+    def __init__(
+        self,
+        keywords: Optional[Iterable] = None,
+        filter_leading_underscore: bool = True,
+    ):
+        keywords = list(keywords) if keywords is not None else []
+        self.filter_leading_underscore = filter_leading_underscore
+        self.keywords = (
+            [keyword for keyword in keywords if keyword[:1] != "_"]
+            if filter_leading_underscore
+            else keywords
+        )
+        self.shortcut: dict[str, str | int] = {}
+        self.abbreviation_dict = defaultdict(list)
+
+        for keyword in self.keywords:
+            self.optimize_symbols(keyword)
+
+        self._rebuild_finalize()
+
+    def __contains__(self, keyword: str) -> bool:
+        return keyword in self.shortcut
+
+    def __getitem__(self, keyword: str) -> Optional[int | str]:
+        return self.shortcut.get(keyword)
+
+    def __delitem__(self, keyword: str) -> None:
+        self.keywords.remove(keyword)
+        self.rebuild()
+
+    def make_abbreviation(self, s: str, groups_length: int) -> str:
```

Comment:
Is this part of the public API (I didn't see any tests or usages in the codebase that uses this method explicitly)? If it's just meant to be internally used by `Shortcut`, can you add a leading underscore. I appreciate the examples here; also can you add a brief sentence or two on what this is achieving?

Reply:
done